icefoganalytics/sfa-client
# Enhancement: Add Letter Creation Pipeline Service

Fixes https://github.com/icefoganalytics/sfa-client/issues/59

# Context

The current letter creation service operates with a funding request id and a letter slug. Now we need a POST endpoint that can generate all appropriate letters for a given funding request id, save them - somewhere to be decided - and return a 201 status.

## Examples Of Pipelines

1. Yukon Grant pipeline needs to generate a student and institution approval letter, or a rejection letter.
2. Student Training Allowance pipeline needs to generate a Yukon University letter, Alkan Air letter, or a rejection letter.
> There are 5 cases in total.

## Implementation

Add an admin funding requests letters CreateService that has a large if block that choses which letter pipeline to run.
Each letter pipeline creates 1 or more letters and uploads them.

## Questions

- [ ] Should I delete all the old funding letter request code, its only useful if people are selecting letters to print via the UI, otherwise its just taking up space.

# Testing Instructions

1. Boot the api and db services via `dev boot`
2. Boot the web service via `cd src/web && npm run start`
4. Generate an appropriate dataset for each case ... ???
5. Check that the new Student Training Allowance Rejection completes correctly at POST http://localhost:3000/api/v2/admin/funding-requests/1122/letters/student-training-allowance-rejection.pdf
6. Check that Yukon Grant Student/Institution completes corrects via POST to http://localhost:3000/api/v2/admin/funding-requests/69473/letters
7. Check that the Yukon Grant Rejection letter completes via POST to http://localhost:3000/api/v2/admin/funding-requests/265/letters
8. Check that the Student Training Allowance Yukon University letter completes correctly; no valid data configuration at this point.
9. Check that the Student Training Allowance Alkan Air letter completes correctly; no valid data configuration at this point.
10. Verify that the appropriate letters were saved to S3, somehow?
